File: sdks/python/src/oris_sdk/signing.py
import base64
import json

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey


def _private_key(seed: bytes) -> Ed25519PrivateKey:
    if len(seed) != 32:
        raise ValueError("seed must be 32 bytes")
    return Ed25519PrivateKey.from_private_bytes(seed)


def sign_body(seed: bytes, body: bytes) -> str:
    key = _private_key(seed)
    sig = key.sign(body)
    return base64.b64encode(sig).decode()


def sign_payload(seed: bytes, payload: object) -> str:
    data = json.dumps(payload, separators=(",", ":"), ensure_ascii=False).encode()
    return sign_body(seed, data)


def public_key_base64(seed: bytes) -> str:
    key = _private_key(seed)
    pub_bytes = key.public_key().public_bytes_raw()
    return base64.b64encode(pub_bytes).decode()


def public_key_hex(seed: bytes) -> str:
    key = _private_key(seed)
    pub_bytes = key.public_key().public_bytes_raw()
    return pub_bytes.hex()

File: sdks/go/internal/signing.go
package internal

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
)

func SignBody(seed [32]byte, body []byte) string {
	key := ed25519.NewKeyFromSeed(seed[:])
	sig := ed25519.Sign(key, body)
	return base64.StdEncoding.EncodeToString(sig)
}

func SignPayload(seed [32]byte, payload interface{}) (string, error) {
	data, err := json.Marshal(payload)
	if err != nil {
		return "", err
	}
	return SignBody(seed, data), nil
}

func PublicKeyBase64(seed [32]byte) string {
	key := ed25519.NewKeyFromSeed(seed[:])
	pub := key.Public().(ed25519.PublicKey)
	return base64.StdEncoding.EncodeToString(pub)
}

func PublicKeyHex(seed [32]byte) string {
	key := ed25519.NewKeyFromSeed(seed[:])
	pub := key.Public().(ed25519.PublicKey)
	return hex.EncodeToString(pub)
}
